openssl-dynamic/src/main/c/ocsp.c

```diff
@@ -0,0 +1,352 @@
+    }
+
+    //
+    // ATTENTION: We need to create a global reference of the callback.
+    //
+    jweak callbackRef = (*e)->NewWeakGlobalRef(e, callback);
+
+    error = SSL_CTX_set_tlsext_status_arg(c->ctx, callbackRef);
+    if (error <= 0L) {
+        openssl_netty_tcnative_ocsp_free_callback(e, NULL, callbackRef);
+
+        tcn_ThrowException(e, "SSL_CTX_set_tlsext_status_arg() failed");
+        return NULL;
+    }
+
+    // The client and server callbacks use slightly different return
+    // values. We need therefore two different functions. We could use
+    // a struct as an argument for SSL_CTX_set_tlsext_status_arg() as
+    // well but that involve a lot more work.
+    error = SSL_CTX_set_tlsext_status_cb(c->ctx,
+        (client ? openssl_netty_tcnative_ocsp_client_callback : openssl_netty_tcnative_ocsp_server_callback));
+
+    if (error <= 0L) {
+        openssl_netty_tcnative_ocsp_free_callback(e, NULL, callbackRef);
+
+        tcn_ThrowException(e, "SSL_CTX_set_tlsext_status_cb() failed");
+        return NULL;
+    }
+
+    return callbackRef;
+#endif
+}
+
+/**
+ * Releases the OCSP callback.
+ */
+TCN_IMPLEMENT_CALL(void, SSLContext, freeOcspCallback)(TCN_STDARGS, jlong ctx, jobject callback)
+{
+#if defined(TCN_NO_OCSP)
+    return;
+#elif defined(OPENSSL_IS_BORINGSSL)
+    // TODO: BoringSSL has OCSP support but uses a different API.
+    return;
+#else
+    tcn_ssl_ctxt_t *c = J2P(ctx, tcn_ssl_ctxt_t *);
+    openssl_netty_tcnative_ocsp_free_callback(e, c, callback);
+#endif
+}
+
+/**
+ * Enables or disables OCSP stapling for the given SSLEngine.
+ */
+TCN_IMPLEMENT_CALL(jboolean, SSL, setStatusType)(TCN_STDARGS, jlong ssl, jint type)
+{
+#if defined(TCN_NO_OCSP)
+    return JNI_FALSE;
+#elif defined(OPENSSL_IS_BORINGSSL)
+    // TODO: BoringSSL has OCSP support but uses a different API.
+    return JNI_FALSE;
+#else
+    SSL *ssl_ = J2P(ssl, SSL *);
+    if (ssl_ == NULL) {
+        tcn_ThrowException(e, "ssl is null");
+        return JNI_FALSE;
+    }
+
+    return (SSL_set_tlsext_status_type(ssl_, type) == 1L) ? JNI_TRUE : JNI_FALSE;
+#endif
+}
+
+/**
+ * Server: Sets the stapled OCSP record bytes for the given SSLEngine.
+ */
+TCN_IMPLEMENT_CALL(jboolean, SSL, setOcspStaple)(TCN_STDARGS, jlong ssl, jbyteArray response)
+{
+#if defined(TCN_NO_OCSP)
+    return JNI_FALSE;
+#elif defined(OPENSSL_IS_BORINGSSL)
+    // TODO: BoringSSL has OCSP support but uses a different API.
+    return JNI_FALSE;
+#else
+    SSL *ssl_ = J2P(ssl, SSL *);
+    if (ssl_ == NULL) {
+        tcn_ThrowException(e, "ssl is null");
+        return JNI_FALSE;
+    }
+
+    jsize length = (*e)->GetArrayLength(e, response);
+    if (length == 0) {
+        return JNI_FALSE;
+    }
+
+    //
+    // ATTENTION: This took a while to figure out but OpenSSL wants to (and will)
+    // free() this pointer on its own. Give it something it can free or it will crash.
+    //
+    unsigned char *value = OPENSSL_malloc(sizeof(unsigned char) * length);
```

Comment:
IIRC for c89 variables have to be declared at the beginning of a new scope ... maybe we use c99 now ... or rely on a GNU extension? @normanmaurer - do you recall?